### PR TITLE
Refactor: Use complex type as the template argument for PW calculation

### PR DESCRIPTION
**Pull request by denghuilu**
As mentioned by #2951 , the usage of template parameters in ABACUS is relatively confusing. In this PR, we employ a unified numerical type template represented by 'T' instead of 'FPTYPE' to denote the calculation's numerical type. This 'T' template is designed to accommodate various data types, including 'std::complex' and others.

### CODE REVIEW

**Review by maki49 — APPROVED** [review]
I will do some LCAO-refactor after this PR is merged.

File: source/module_hamilt_general/hamilt.h [hunk 1]
```diff
@@ -33,7 +33,7 @@ class Hamilt
     int non_first_scf=0;
 
     // first node operator, add operations from each operators
-    Operator<std::complex<FPTYPE>, Device>* ops = nullptr;
+    Operator<T, Device>* ops = nullptr;
```
Comment (dyzheng) [inline source/module_hamilt_general/hamilt.h:36]: When T=double, ops and opds will be same

Comment (maki49) [reply source/module_hamilt_general/hamilt.h:36]: Sure. The next step I want to do is to unify opsd into ops.
I think there should be only one `Operator` instance in class `Hamilt`.

Comment (denghuilu) [reply source/module_hamilt_general/hamilt.h:36]: T can not be double right now(This requires a lot of subsequent code changes, including updating the related classes and the underlying multi-device ops.) However, if we support T taking double in the future, can the opds be removed? Is it necessary to have both ops and opds in the basic class at the same time?

File: source/module_elecstate/elecstate_pw.cpp [hunk 2]
```diff
@@ -29,38 +29,38 @@ ElecStatePW<FPTYPE, Device>::~ElecStatePW()
...
 }
 
-template<typename FPTYPE, typename Device>
-void ElecStatePW<FPTYPE, Device>::init_rho_data() 
+template<typename T, typename Device>
+void ElecStatePW<T, Device>::init_rho_data() 
 {
     if (GlobalV::device_flag == "gpu" || GlobalV::precision_flag == "single") {
-        this->rho = new FPTYPE*[this->charge->nspin];
+        this->rho = new R*[this->charge->nspin];
```
Comment (wenfei-li) [inline source/module_elecstate/elecstate_pw.cpp]: What is this 'R' here? Is it a bug or intended?

Comment (denghuilu) [reply source/module_elecstate/elecstate_pw.cpp]: `R` represents `Real type` here. The value of `T` is consistent with the template parameter of `Psi`, which is a complex number type currently, such as `complex<double>`. In this case, there are still some data that are real number types, such as `double`, so the `R` here represents the real number type. , more specifically their conversion relationship is as follows:
```
template <typename T>
struct PossibleComplexToReal {
    using type = T; /**< The return type based on the input type. */
};

/**
 * @brief Specialization of PossibleComplexToReal for std::complex<float>.
 *
 * This specialization sets the return type to be float when the input type is std::complex<float>.
 */
template <>
struct PossibleComplexToReal<std::complex<float>> {
    using type = float; /**< The return type specialization for std::complex<float>. */
};

/**
 * @brief Specialization of PossibleComplexToReal for std::complex<double>.
 *
 * This specialization sets the return type to be double when the input type is std::complex<double>.
 */
template <>
struct PossibleComplexToReal<std::complex<double>> {
    using type = double; /**< The return type specialization for std::complex<double>. */
};
```
```
template<typename T, typename Device = psi::DEVICE_CPU>
class ElecStatePW : public ElecState
{
  private:
    using R = typename PossibleComplexToReal<T>::type;
    ...
};
```

Comment (dyzheng) [reply source/module_elecstate/elecstate_pw.cpp]: @denghuilu I think this is good design for dealing with complex<real>/real , but I think the name of R and PossibleComplexToReal are a little bit confusing, do you mind to rename to "Real" and "GetTypeReal"

Comment (denghuilu) [reply source/module_elecstate/elecstate_pw.cpp]: Sure, I'll take care of it.

File: source/module_hamilt_general/hamilt.h (hunk 1 above)
Comment (dyzheng) [reply source/module_hamilt_general/hamilt.h:36]: > Sure. The next step I want to do is to unify opsd into ops. I think there should be only one `Operator` instance in class `Hamilt`.

Okay, I think so.

**Review by dyzheng — APPROVED** [review]

**Pull request merged**